pkg/overlay/cache.go

```diff
@@ -107,51 +109,67 @@ func (o *Cache) Bootstrap(ctx context.Context) error {
 			return err
 		}
 	}
-	// called after kademlia is bootstrapped
-	// needs to take RoutingTable and start to persist it into the cache
-	// take bootstrap node
-	// get their route table
-	// loop through nodes in RT and get THEIR route table
-	// keep going forever and ever
-
-	// Other Possibilities: Randomly generate node ID's to ask for?
-
-	_, err = o.DHT.GetRoutingTable(ctx)
 
 	return err
 }
 
-// Refresh walks the network looking for new nodes and pings existing nodes to eliminate stale addresses
+// Refresh updates the cache db with the current DHT.
+// We currently do not penalize nodes that are unresponsive,
+// but should in the future.
 func (o *Cache) Refresh(ctx context.Context) error {
-	// iterate over all nodes
-	// compare responses to find new nodes
-	// listen for responses from existing nodes
-	// if no response from existing, then mark it as offline for time period
-	// if responds, it refreshes in DHT
-	_, rtErr := o.DHT.GetRoutingTable(ctx)
-
-	if rtErr != nil {
-		return rtErr
+	log.Print("starting cache refresh")
+	table, err := o.DHT.GetRoutingTable(ctx)
+	if err != nil {
+		zap.Error(OverlayError.New("Error getting routing table", err))
+		return err
 	}
 
-	_, err := o.DHT.GetNodes(ctx, "0", 128)
+	k := table.K()
+	buckets, err := table.GetBuckets()
+	fmt.Printf("got buckets %+v\n, error %s", buckets, err)
+	nodes, err := o.DHT.GetNodes(ctx, "", k)
 
-	if err != nil {
-		return err
+	/// kick off concurrent refreshes
+	go func() {
+		randomId, err := kademlia.NewID()
+		if err != nil {
+			zap.Error(OverlayError.New("Error generating random ID", err))
+			fmt.Printf("error finding random ID", err)
+		}
+
+		ping, err := o.DHT.Ping(ctx, overlay.Node{Id: randomId.String()})
+		if err != nil {
+			zap.Error(OverlayError.New("Error pinging overlay node", err))
+			fmt.Printf("error pinging node %+v\n", overlay.Node{Id: randomId.String()})
+		}
+		o.DB.Put([]byte(ping.Id), []byte(ping.Address.Address))
+	}()
+
+	for _, node := range nodes {
+		pinged, err := o.DHT.Ping(ctx, *node)
+		if err != nil {
+			zap.Error(ErrNodeNotFound)
+		} else {
+			o.DB.Put([]byte(pinged.Id), []byte(pinged.Address.Address))
+		}
 	}
 
-	return nil
+	return err
 }
 
-// Walk iterates over buckets to traverse the network
+// Walk iterates over each node in each bucket to traverse the network
 func (o *Cache) Walk(ctx context.Context) error {
-	nodes, err := o.DHT.GetNodes(ctx, "0", 128)
+	table, _ := o.DHT.GetRoutingTable(ctx)
+	k := table.K()
+
```

Comment:
Do we want to get all the nodes in the table, or just the K nearest nodes? And why are we starting with "0"? I'm just trying to understand what this function is supposed to do :)

Reply:
I'm actually going to change it to be `""` because that will use the nodes Self ID as the starting point, which makes more sense. I am going to expand out the limit on this as well, so that we can get more nodes.